src/controller/InteractionPeer.cpp

```diff
@@ -0,0 +1,121 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <controller/InteractionPeer.h>
+
+namespace chip {
+namespace Controller {
+
+CHIP_ERROR InteractionPeer::SendMessage(Protocols::Id protocolId, uint8_t msgType, Messaging::SendFlags sendFlags,
+                                        System::PacketBufferHandle && message)
+{
+    return mDelegate->SendMessage(protocolId, msgType, sendFlags, std::move(message));
+}
+
+CHIP_ERROR InteractionPeer::SendReadAttributeRequest(app::AttributePathParams aPath, Callback::Cancelable * onSuccessCallback,
+                                                     Callback::Cancelable * onFailureCallback, app::TLVDataFilter aTlvDataFilter)
+{
+
+    uint8_t seqNum = GetNextSequenceNumber();
+    aPath.mNodeId  = mDelegate->GetDeviceId();
+
+    ReturnErrorOnFailure(mDelegate->PreparePeer());
+
+    if (onSuccessCallback != nullptr || onFailureCallback != nullptr)
+    {
+        AddResponseHandler(seqNum, onSuccessCallback, onFailureCallback, aTlvDataFilter);
+    }
+    // The application context is used to identify different requests from client applicaiton the type of it is intptr_t, here we
+    // use the seqNum.
+    CHIP_ERROR err = chip::app::InteractionModelEngine::GetInstance()->SendReadRequest(
+        mDelegate->GetDeviceId(), 0, mDelegate->GetSesssionHandle(), nullptr /*event path params list*/, 0, &aPath, 1,
+        0 /* event number */, seqNum /* application context */);
+    if (err != CHIP_NO_ERROR)
+    {
+        CancelResponseHandler(seqNum);
+    }
+    return err;
+}
+
+CHIP_ERROR InteractionPeer::SendWriteAttributeRequest(app::WriteClientHandle aHandle, Callback::Cancelable * onSuccessCallback,
+                                                      Callback::Cancelable * onFailureCallback)
+{
+    uint8_t seqNum = GetNextSequenceNumber();
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    aHandle->SetAppIdentifier(seqNum);
+    ReturnErrorOnFailure(mDelegate->PreparePeer());
+
+    if (onSuccessCallback != nullptr || onFailureCallback != nullptr)
+    {
+        AddResponseHandler(seqNum, onSuccessCallback, onFailureCallback);
```

Comment:
If AddResponseHandler supports nullptr in either onSuccess/onFailureCallback, the outer `if` is not needed.